clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java

```diff
@@ -0,0 +1,198 @@
+    }
+    if (heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()) == null) {
+      heartbeatTimestamps.get(version.getStoreName()).put(version.getNumber(), new VeniceConcurrentHashMap<>());
+    }
+    if (heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()).get(partition) == null) {
+      Map<String, Long> regionTimestamps = new VeniceConcurrentHashMap<>();
+      if (version.isActiveActiveReplicationEnabled()) {
+        for (String region: regionNames) {
+          regionTimestamps.put(region, DEFAULT_SENTINEL_HEARTBEAT_TIMESTAMP);
+        }
+      } else {
+        regionTimestamps.put(localRegionName, DEFAULT_SENTINEL_HEARTBEAT_TIMESTAMP);
+      }
+      heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()).put(partition, regionTimestamps);
+    }
+  }
+
+  private synchronized void removeEntry(
+      Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> heartbeatTimestamps,
+      Version version,
+      int partition) {
+    if (heartbeatTimestamps.get(version.getStoreName()) != null) {
+      if (heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()) != null) {
+        if (heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()).get(partition) != null) {
+          heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()).remove(partition);
+        }
+      }
+    }
+  }
+
+  public void addFollowerLagMonitor(Version version, int partition) {
+    initializeEntry(followerHeartbeatTimeStamps, version, partition);
+    removeEntry(leaderHeartbeatTimeStamps, version, partition);
+  }
+
+  public void addLeaderLagMonitor(Version version, int partition) {
+    initializeEntry(leaderHeartbeatTimeStamps, version, partition);
+    removeEntry(followerHeartbeatTimeStamps, version, partition);
+  }
+
+  public void removeLagMonitor(Version version, int partition) {
+    removeEntry(leaderHeartbeatTimeStamps, version, partition);
+    removeEntry(followerHeartbeatTimeStamps, version, partition);
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    reportingThread.start();
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    reportingThread.interrupt();
+  }
+
+  public void recordLeaderHeartbeat(String store, int version, int partition, String region, Long timestamp) {
+    recordHeartbeat(store, version, partition, region, timestamp, leaderHeartbeatTimeStamps);
+
+  }
+
+  public void recordFollowerHeartbeat(String store, int version, int partition, String region, Long timestamp) {
+    recordHeartbeat(store, version, partition, region, timestamp, followerHeartbeatTimeStamps);
+  }
+
+  private void recordHeartbeat(
+      String store,
+      int version,
+      int partition,
+      String region,
+      Long timestamp,
+      Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> heartbeatTimestamps) {
+    if (heartbeatTimestamps.get(store) != null) {
```

Comment:
Should we emit any metric/logs if entries are missing? i.e. `removeLagMonitor` is called on the corresponding store partition but somehow the ingestion is still triggering `recordHeartbeat` (maybe when draining the drainer queue?).

Reply:
That's kind of interesting 🤔 

So like... as a means to detect leaked ingestion?  I think that's not a bad idea, but maybe let's do that in a separate PR and focus on this problem we're trying to detect.

Reply:
Sounds good.